### ISSUE TITLE
How to secure JSF pages using FORM authentication

### ISSUE THREAD

**Issue by ratcashdev** [javaee-security-spec/soteria#9]
A bit of guidance needed for the intended usage for FORM based authentication, as this is not clear from the test cases.

Testing on m02-SNAPSHOT (current master branch) on Payara 4.1.1.162

Suppose we have a JSF war.  The `{context}/auth/signin.xhtml` submits to `{context}/auth/j_security_check`.
To setup Authentication, I did the following

```
@FormAuthenticationMechanismDefinition(
    loginToContinue = @LoginToContinue(
        loginPage="auth/signin",
        errorPage="",useForwardToLogin = true
    )
)
@ApplicationScoped
public class SoteriaSetup  {
// empty class
}
```

In addition, i created my own IdentityStore:

```
@RequestScoped
public class MyCredentialStore implements IdentityStore { ... }
```

in web.xml I set up the constraints:

```
<security-constraint>
        <display-name>Secure pages</display-name>
        <web-resource-collection>
            <web-resource-name>UI Resource</web-resource-name>
            <description>UI</description>
            <url-pattern>/restricted/*</url-pattern>
            <http-method>GET</http-method>
            <http-method>POST</http-method>
        </web-resource-collection>
        <auth-constraint>
            <description>Only for users</description>
            <role-name>users</role-name>
        </auth-constraint>
    </security-constraint>
```

There's no login-config section in the web.xml. The group to role mapping is also done in glassfish-web.xml.

All seems to be OK (the custom code in `MyCredentialStore` gets called and returns Success). However:
1. the login attempt ends with HTTP 404 for the url `..../j_security_check`.
2. I receive a HTTP 500 (without any trace in the server logs) when I try to directly access a valid protected resource while not yet authenticated.
3. Is there a way to tell SOTERIA where to go after a successful login attempt, if the originally intended landing page was directly the login page, e.g.  '/auth/signin.xhtml' ?
4. It would be nice if the authenticator also recognized `.../j_security_exit` and did the necessary cleanup for a proper logout.

BTW: some attempts resulted in HTTP 401 for point 1 and 2 above because the container wanted to do a "BASIC" authentication with the error `Warning:   WEB9102: Web Login Failed: com.sun.enterprise.security.auth.login.common.LoginException: Login failed: Failed file login for [username]` if it failed. Not sure why it happened. Maybe configuration changes between hot redeployments are not handled gracefully?

**Comment by arjantijms** [comment]
For usage with JSF the custom form mechanism is often the best one. This lets you use a form page that's more of a "real" JSF page. This means you can use validation and set a Faces message etc. 

See https://github.com/javaee-security-spec/soteria/tree/master/test/app-mem-customform

From your description, it's likely better if a backing bean behind /auth/signin.xhtml calls `securityContext.authenticate` and then sets the` newAuthentication(true)` parameter. I noticed the test I wrote for this somehow didn't make it into the repo (I probably forgot to commit it). I'll try to add it soon.

Nevertheless, the classic FORM one (represented in Soteria by `@ FormAuthenticationMechanismDefinition`) should theoretically work as well in JSF, but with some caveats as described below:

>Suppose we have a JSF war. The {context}/auth/signin.xhtml submits to {context}/auth/j_security_check

This is probably the issue. The regular FORM mechanism is designed to be a login-to-continue type. It does not (currently) support starting a new (explicitly triggered) authentication.

I'm not sure if the classic FORM authentication mechanism as supplied by the Servlet spec is guaranteed to support this either. It may work on some containers, but I don't think it's in the spec.

What I think happens now is that "/auth/j_security_check" is the initial request (for the authentication chain), that does kickoff authentication though. After having successfully authenticated it goes on to invoke the original URL (/auth/j_security_check), which of course doesn't exists, so results in a 404.

The question is if FORM should indeed support this, as we tried to keep the Soteria supplied FORM close to the Servlet one.

In general, the difference between a "new" authentication and a "login-to-continue" is very ill understood. See also this discussion: https://java.net/projects/servlet-spec/lists/users/archive/2016-04/message/35


>2. I receive a HTTP 500 (without any trace in the server logs) when I try to directly access a valid protected resource while not yet authenticated.

Does this happen after you tried to authenticate explicitly by requesting /auth/j_security_check? Could be that the state that's kept by login-to-continue gets in the way here.

>3. Is there a way to tell SOTERIA where to go after a successful login attempt, if the originally intended landing page was directly the login page, e.g. '/auth/signin.xhtml' ?

For the login-to-continue flow, it's of course the originally requested page. Explicit authentication via /j_security_check is not (yet) supported. I'm wary to support this via heuristics, as the flow and state bookkeeping is already hideously complex. Maybe either a different URL like /j_security_check_new, or a request parameter like /j_security_check?new=true would be the safer option here.

For an explicit (new) authentication via `securityContext.authenticate` I've indeed been experimenting with letting the user provide a URL to redirect to after a successful authentication.

See: 

https://github.com/javaee-security-spec/soteria/blob/master/api/src/main/java/javax/security/authentication/mechanism/http/AuthenticationParameters.java#L64

and

https://github.com/javaee-security-spec/soteria/blob/master/impl/src/main/java/org/glassfish/soteria/cdi/LoginToContinueInterceptor.java#L161

But this has not been implemented yet as of m02.

>4. It would be nice if the authenticator also recognized .../j_security_exit and did the necessary cleanup for a proper logout.

Principally speaking this could be implemented. The idea is now however that the application would call `request.logout` or (when it's ready) `securityContext.logout`. You could however propose extending the classic FORM mechanism with several additional "URL commands" like /j_security_exit at the JSR 375 user mailing list. If there's enough support I'm sure we can implement it.

>BTW: some attempts resulted in HTTP 401 for point 1 and 2 above because the container wanted to do a "BASIC" authentication with the error Warning: WEB9102: Web Login Failed: com.sun.enterprise.security.auth.login.common.LoginException: Login failed: Failed file login for [username] if it failed.

This is the Servlet FORM authentication mechanism kicking in, and then calling the native GlassFish identity store (which defaults to a file based realm). With a Soteria authentication mechanism present this should not happen, so it might be interesting to see what exactly causes this.

**Comment by ratcashdev** [comment]
Thank you for the very detailed explanation. 
You're certainly right, that for a JSF page a CUSTOM FORM may be more appropriate. I will try to get it work and report back.

>I'm not sure if the classic FORM authentication mechanism as supplied by the Servlet spec is guaranteed to support this either. It may work on some containers, but I don't think it's in the spec.

Apparently I have serious deficiencies in this topic. I based my implementation on the article here:
https://docs.oracle.com/cd/E19226-01/820-7627/bncby/index.html

In that article, the form directly submits to j_security_check, just like in my case. 

I have also realized that to be consistent with the Login-to-continue model, it is still easy to create an explicit login link. I can simply use any page that is inside the restricted area. So far so good.

However, how can I prevent someone from calling the /auth/signin.xhtml directly? If this happens, the login-to-continue breaks...

The previous library (called AuthenticRoast, slightly updated for CDI), had a parameter to specify "where to go next, if destination was not explicitly given". How much this one conforms to the standards I have no idea, but I considered it a clever solution to the problem above.

>Does this happen after you tried to authenticate explicitly by requesting /auth/j_security_check?

It happens regardless of that state. I wonder if the path `loginPage="auth/signin"` is correct?

**Comment by ratcashdev** [comment]
I have read https://java.net/projects/servlet-spec/lists/users/archive/2016-04/message/35

I think I understand the problem. In the very last comment you mention OAUTH posting back to /login.
I think this would not be a very good solution. I'd rather provide a different virtual callback url, e.g. `/social/google/callback`. The Interceptor would spot this and handle appropriately similarly to how it captures a post to `j_security_check`.

Virtual callbacks are also, imho, the solution to the problem. Using them you can clearly tell the authentication mechanism what is the state, and from which state to continue or start new. Example:
```
/auth/signin -> /auth/social/facebook/start
/auth/social/facebook -> www.facebook.com/login?paramsHere
www.facebook.com -> /auth/social/facebook/callback?paramsHere
```
and so on. Thus as long as you are in control of the URL, you're good. Using this you don't even need the request.authenticate() method. Just simply post to the URL of your choice. Obviously, your `HttpAuthenticationMechanism` will have to know which URLs to intercept and how to treat them.

This takes me to an other  question. How to combine multiple Authentication Mechanism?
One way would be to write MyCustomHttpAuthenticationMechanism that would inspect the URL and call the respective mechanism. However, i don't see how in this case can I re-use the oh-so-convenient annotations, like `@FormAuthenticationMechanismDefinition`

**Comment by arjantijms** [comment]
>I think I understand the problem. In the very last comment you mention OAUTH posting back to /login.
I think this would not be a very good solution. I'd rather provide a different virtual callback url, e.g. /social/google/callback.

Sure, it's probably better to use a different URL there. In that example and in the documentation for the example OAuth provider for Soteria (see https://github.com/omnifaces/soteria-google-oauth-client) I just happened to use the same URL. But it can be different ones.

>This takes me to an other question. How to combine multiple Authentication Mechanism?

This is an open topic. Darran Lofthouse had an elaborate proposal for this. See https://java.net/jira/browse/JAVAEE_SECURITY_SPEC-37

>One way would be to write MyCustomHttpAuthenticationMechanism that would inspect the URL and call the respective mechanism. However, i don't see how in this case can I re-use the oh-so-convenient annotations, like @FormAuthenticationMechanismDefinition

This is also an open topic indeed (we have many open topics it seems). The `@*AuthenticationMechanismDefinition` annotations are essentially a configuration mechanism that make a configured bean available for injection to CDI. We've been exploring some ways to make multiple of such beans available. Candidates include having a `name` or `qualifier` attribute. With that in place you can either inject a MyCustomHttpAuthenticationMechanism with different mechanisms using qualifiers, or do a programmatic lookup via either the name or type + qualifier.

A bit offtopic, but one thing we're particularly struggling with is adding interceptor annotations like `@RememberMe` to a mechanism introduced via `@*AuthenticationMechanismDefinition`. In CDI 1.2 this is completely non-trivial/impossible. Maybe CDI 2.0 will make this possible.

>The point is, if those URLs were user-pluggable, then we don't have to standardize them.

True, they could just as well be given via attributes on the `@FormAuthenticationMechanismDefinition`. The EG discussion would be in the first place for having the additional "action urls" at all, and then could be followed-up by making the names configurable. I'm not sure why the Servlet spec never made the URL configurable btw.

**Comment by ratcashdev** [comment]
BTW: got the problem for point 2 (HTTP 500):
```
Exception occurred in target VM: setAttribute: Non-serializable attribute with name org.glassfish.soteria.original.request 
java.lang.IllegalArgumentException: setAttribute: Non-serializable attribute with name org.glassfish.soteria.original.request
	at org.apache.catalina.session.ManagerBase.checkSessionAttribute(ManagerBase.java:835)
	at org.apache.catalina.session.StandardSession.setAttribute(StandardSession.java:1841)
	at org.apache.catalina.session.StandardSessionFacade.setAttribute(StandardSessionFacade.java:178)
	at org.glassfish.soteria.cdi.LoginToContinueInterceptor.saveRequest(LoginToContinueInterceptor.java:364)
	at org.glassfish.soteria.cdi.LoginToContinueInterceptor.processContainerInitiatedAuthentication(LoginToContinueInterceptor.java:180)
	at org.glassfish.soteria.cdi.LoginToContinueInterceptor.validateRequest(LoginToContinueInterceptor.java:122)

```
Easy fix for this: make RequestData Serializable. This fixes issue 2 in the original topic.


Now to get issue 1 working, we need to make `CredentialValidationResult` and `CallerPrincipal` also Serializable
```
Exception occurred in target VM: setAttribute: Non-serializable attribute with name org.glassfish.soteria.authentication 
java.lang.IllegalArgumentException: setAttribute: Non-serializable attribute with name org.glassfish.soteria.authentication
	at org.apache.catalina.session.ManagerBase.checkSessionAttribute(ManagerBase.java:835)
	at org.apache.catalina.session.StandardSession.setAttribute(StandardSession.java:1841)
	at org.apache.catalina.session.StandardSessionFacade.setAttribute(StandardSessionFacade.java:178)
	at org.glassfish.soteria.cdi.LoginToContinueInterceptor.saveAuthentication(LoginToContinueInterceptor.java:385)
	at
```

However, this is not enough. Is the UnmodifiableList (used for group storage) serializable?

BTW: Not sure why it needs to be serializable. I am not running in a cluster and the web.xml does not contain the `distributable` tag.

**Comment by arjantijms** [comment]
Thanks a lot for reporting this! I'll take a look at it.

>Not sure why it needs to be serializable. I am not running in a cluster and the web.xml does not contain the distributable tag.

Officially everything you put in the session has to be serializable. Sonar checks for it as well, but I've yet to walk through all Sonar warnings. The resources available for Soteria are quite limited.

**Comment by ratcashdev** [comment]
>Officially everything you put in the session has to be serializable.

My point was that Payara does not complain for the non-serializable classes when running the test cases, but it does so when using soteria in my own web-app.

>..., how can I prevent someone from calling the /auth/signin.xhtml directly? If he/she does, the login-to-continue breaks...
Any ideas how to solve this in the current version of soteria?

**Comment by arjantijms** [comment]
> point was that Payara does not complain for the non-serializable classes when running the test cases, but it does so when using soteria in my own web-app.

I understand what you mean. I'm using Soteria internally for a couple of (test) applications but haven't seen the issue there either, so it's not just the tests. Could it be that you have preserve session on restart activated?

In OmniFaces we saw a comparable case, where we had some stuff in the session which never failed and then in some cases suddenly it did fail.

>Any ideas how to solve this with the current version of Soteria?

For the moment you could check the "org.glassfish.soteria.original.request" key in the session. If the corresponding value is non-null there's a remembered request and you're in the middle of an authentication dialog.

We have to see how this can best be solved going forward. As I mentioned, we have very few resources (man hours) available to work on Soteria. Your comments help a lot though, so thanks a lot for that!

**Comment by ratcashdev** [comment]
slightly off topic, now that the problem with using FORM is now resolved!

> I'm wary to support this via heuristics, as the flow and state bookkeeping is already hideously complex.

If I may (and a lot of apologies in advance, as I am not an EG member, and chances are you've already discussed such topics long ago), I'd not do this. Rather keep thing simple.
 
Assuming the scope of JSR375 (and soteria) are HTTP[S]-only requests (could not really find this info in the JSR) and we can forget the rest (e.g. securing protocols for Remote EJB calls), imo, for state management, the framework itself should only deal with the following:

1. store desired destination if :
 * url is not one of the virtual URLs (like the login page, or "j_security_check") understood by the pluggable mechanisms
 * url is restricted
 * request is not yet authenticated (by calling the pluggable mechanism [chains])
2. Attempt the authentication using the plugged mechanisms
 * a) authenticate the request with the provided credentials if those are available.
 * b) trigger the authentication using the plugged mechanisms
3. continue to original destination when authentication succeeded
 * I've noticed Soteria copies the request and re-aplies the credentials for the final destination to avoid a new authentication cycle. This may or may not work, depending on the underlying mechanism. I would not do it. Instead the framework should be prepared to authenticate every request separately. It's the mechanisms job to optimize (by the means of a "RememberMe" cookie, or session attribute, for example) this.
 * if the destination is not known (direct access to the login page, for example), ask the mechanism for the default destination

The state during phase 2 - if at all needed - should be handled by the specific authentication module (e.g. HTTP, Kerberos, OAuth*, SSL). The already defined touch-points (FAILED, CONTINUE, SUCCESS, according to JSR196) with the framework seem to be sufficient to me.

Corner cases:
1. If the authentication conversation breaks during phase 2 (e.g. the user leaves the login page) and then returns to a different protected resource, the framework should start with phase 1 and reset the destination.
2. If the authentication conversation breaks during phase 2 (e.g. the user leaves the login page) and then returns to any out-of-order state in the mechanism (and chances are, some important data are missing), the state within the pluggable mechanism should be reset and start from the very first step again (the login page)
3. Application code calls the request.authenticate() method. This is a tough one. Take, for example, a 2 factor authentication model, where application code collects the information, but (application provided) container code does the verification for every stage separately.

My initial thought is that request.authenticate() should simply trigger phase 1 and 2 in the framework. Next, the mechanism will provide the necessary challenge response (header, or redirection to a form). When the form is filled in (username/pw or one time code) it posts somewhere. If it posts back to the application for some extra processing, then indeed, we need to have a way to communicate with the mechanism via an API (SecurityContext, perhaps). If the form POSTs to (or GETs) a virtual URL handled by the mechanism, things are way simpler. Anything else could become a fairly complex scenario because of the differences of the underlying mechanisms. This latter case handles better cases, when the response to the challenge is provided directly by the browser's native code (e.g. Kerberos)

Optionally, the framework should also allow us building hierarchies of authentication mechanisms using a declarative approach. While one can always build the hierarchy manually from stand-alone mechanisms, I presume 90% of the cases out there will benefit from the chaining of: "check rememberMe cookie", then "check sessionCookie", and only "trigger expensive authentication" as a last resort. The (small) problem with interceptors are that their order (AFAIK) is only defined in the beans.xml

By hierarchy I mean allowing the mechanisms to:
* operate in parallel for different parts of the application. Here the mechanism should be able to tell the region he is supposed to protect, and this should supersede settings in web.xml. Not sure if this is a good idea though, but I don't have a better one, for now.
* be chained. Example: first check the validity of the session and only if not valid, trigger the expensive authentication (the second mechanism in chain). The order of the mechanisms in this case is very important.

Without a proper hierarchy we could quickly run into endless redirects when different mechanisms would propose to go to a different login-page.

**Comment by arjantijms** [comment]
The login-to-continue algorithm used now is largely that of Servlet 13.6.3, but extracted into a separate interceptor so that it works with pretty much any authentication mechanism.

>Assuming the scope of JSR375 (and soteria) are [Http]ServletRequest(s) (could not really find this info in the JSR) and we can forget the rest (e.g. securing protocols for Remote EJB calls)

That's right. This was discussed, and the outcome was more or less that we indeed focus on the web layer for now. That's why the name of the JSR375 authentication mechanism is `HttpAuthenticationMechanism` (although formally it's still in the proposal state and it could still change). 

More strictly speaking, the `HttpAuthenticationMechanism` introduces specific types for the Servlet Container Profile of JASPIC. We also debated for a (long) time whether we should not add those types directly to JASPIC, but eventually it was (again, more or less) decided to put that type in JSR 375, leaving the lower level abstract types to JASPIC, and thus the higher level more concrete ones to JSR 375.

Note that JASPIC only defines two real profiles, one for Servlet and one for SOAP. With SOAP being so de-emphasised, it's practically speaking only Servlet. (a so-called bridge profile also exists, but it's not about a protocol, but about delegating to a JAAS based identity store).

It was originally planned to have more profiles for JASPIC (to indeed cover remote EJB, but also things like JMS). We briefly discussed whether such profile should now be added to JASPIC, but since remote EJB is also being de-emphasised it was decided against this.

Also note that this is all against the background of very little resources being available to work on either JSR 196 or JSR 375.

>I've noticed Soteria copies the request and re-aplies the credentials for the final destination to avoid a new authentication cycle. Is this really needed?

The important data of the request is copied to make sure the repeated request after authentication is as close to the original request as possible, with all headers and what have you intact. If you only remember the URL, then authenticate, then redirect or forward to that original URL, you'll lose those headers and other data.

If you look at how Servlet containers implement FORM, you'll see them all doing pretty much the same thing.

The authenticated identity is stored once separately, since JASPIC can not authenticate (and remember that authentication) and redirect at the same time. If you actually authenticate (via the identity store), then you can not yet feed the resulting caller name and groups to the JASPIC handler if you still have to redirect (at least not without tricks). See https://java.net/jira/browse/JASPIC_SPEC-23

>The state during phase 2 - if needed - should be handled directly by the specific authentication module (e.g. HTTP, Kerberos, OAuth*, SSL).

This is indeed what happens now. The `LoginToContinueInterceptor` class basically only keeps the state related to the desired destination. The actual authentication mechanism handles its own state. This is most apparent if you use combine the `LoginToContinueInterceptor` with the OAuth authentication mechanism (see the example in the readme here: https://github.com/omnifaces/soteria-google-oauth-client). I had that running for a while in an internal prototype application and it worked pretty well.

>Corner cases:
>1 ...
>2 ...

These are pretty much what makes it complicated, and indeed what makes implementing FORM in plain Servlet containers complicated. If you look at the implementation code of `LoginToContinueInterceptor` you'll see it uses a number of heuristics to determine if any of those cases has occurred. It was thinking though to store the phases more explicitly. I.e. it now does

```java
private boolean isOnLoginPostback(HttpServletRequest request) {
    return 
        getSavedRequest(request) != null &&
        getSavedAuthentication(request) == null;
}
```

Instead of that it may be clearer to put an `org...phase = ON_LOGIN_POSTBACK` or something in the session.

In addition the difference between letting the application continue an authentication dialog and starting a new one from scratch also adds some complexity.

>Note: To keep backward compatibility, i think request.authenticate() should simply trigger phase 1 and 2 in the framework. A separate API (a new method in the SecurityContext perhaps) would be used to forward specific information from the application to the specific mechanisms.

The `SecurityContext#authenticate` draft currently has a facility to convey extra information, most importantly the "newAuthentication" bit. I tried to propose something similar for `request#authenticate` to keep them somewhat in sync, but as you may have read I wasn't quite successful there.

On a related note, the outcome of `request#authenticate` also isn't entirely clear, as different servers return different things there. This is an extra complication.

>Optionally, the framework should also allow us building hierarchies of authentication mechanisms using a simple declarative approach.

There's different things here. Darren Lofthouse from Red Hat proposed multiple actual authentication mechanisms for a single resource, see https://java.net/jira/browse/JAVAEE_SECURITY_SPEC-37

Then there's also an earlier proposal by Alex from Oracle to have multiple authentication mechanisms, but each for its own URL pattern.

I also had a proposal for multiple authentication mechanisms, but where the user could choose one.

And then there's the interceptor chaining (see below). Somehow it will all have to be specified and implemented to work together.

>, I presume 90% of the cases out there will benefit from the chaining of: "check rememberMe cookie", then "check the presence of a Principal in the Session", and only "trigger expensive authentication" as a last resort. 

This is indeed exactly what has been implemented for JSR 375 ;) (do note that check principal in session is the least expensive check so should be the first). I.e. as per the previous OAuth example:

```java
@ApplicationScoped
@AutoApplySession
@RememberMe
@LoginToContinue(
    loginPage="/some_login_page.xhtml"
)
public class OAuthClientAuthenticationMechanism ...
```

>A small issue with using interceptors for this is that their order (AFAIK) can only be defined in the beans.xml. But a sane default in the framework likely solves this for most of the cases.

`@Priority` can also be used. We've given it defaults that are hopefully sane enough:

```java
@Interceptor
@LoginToContinue
@Priority(PLATFORM_BEFORE + 220)
public class LoginToContinueInterceptor ...
```

>Without a proper hierarchy we could quickly run into endless redirects when different mechanisms would propose to go to a different login-page.

Indeed, so that's the difficulty in the overal "multi authentication mechanism" epic.

**Comment by ratcashdev** [comment]
Back again after some time.
I went through https://java.net/projects/servlet-spec/lists/users/archive/2016-04/message/35 once again while trying to implement simple(?) Social Login mechanism.
While I still think that virtual URLs are "a solution", as of now I don't think they're the best solution. Apologies for the initial opposition.

I came across a number of situations where an extended API to talk to the Container mechanism would, indeed, be welcome:
1. User initiates a Social login, which, after redirecting back to the app discovers that the account is not yet created. It needs to send the user to a specific page to ask some more details (and agreement with the registrations, T&Cs, etc.) and then create a record in the underlying identity store. At this point, we have 2 choices:
1a. Application code creates the record in the identity store. However, to make the user logged-in, a Principal must exist, which can only be created by Container's Custom Auth mechanism. we need a way to tell the container that this is to happen (incl. the necessary information/state). A virtual URL is fine (though including the state may be a PITA), but an API would make things more seamless
1b. Container code  creates the identity store record. But for this we need a way to tell the container code that it has to do actually THIS. Again, virtual URL may be some help, but an API would make it seamless
2. After registration we need to mark an email as 'valid'. This is usually done by one-time links sent out in a welcome email by the application. Clicking on that link is either handled by the:
2a. the application, in which case we need a way to tell the container mechanism to create the principal and consider the user as logged-in
2b. by the container code with a specific virtual URL, e.g. /j_security_validEmail?hash=******** which would in turn call back to application code to validate the hash and update the necessary records in the identity store.
3. Two (or multi) factor authentication
